File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    min_time = min(times)
    end = n * min_time
    start = 0 
    answer = 0
    
    while start <= end:
        mid = (start + end) // 2
        
        people = 0
        
        for time in times:
            people += mid // time
        
        if people >= n:
            answer = mid
            end = mid - 1
        
        else:
            start = mid + 1
    
    return answer
        

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        long left = 1;
        long right = (long) Arrays.stream(times).min().getAsInt() * n;
        long answer = 0;
        
        while (left <= right) {
            long mid = left + (right - left) / 2;
            long target = 0;
            
            for (long time : times) {
                target += (mid / time);
            }
            
            if (target >= n) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid +1;
            }
            
        }
        
        return answer;
    }
}